### ISSUE TITLE
ddtrace 1.4.0 not installing in JRuby

### ISSUE THREAD

**Issue by HoneyryderChuck** [DataDog/dd-trace-rb#2244]
Running "bundle install" inside the jruby docker container is somehow trying to install the C extensions.

```
Fetching ddtrace 1.4.0
Installing ddtrace 1.4.0 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.
current directory:
/builds/honeyryderchuck/tobox/vendor/jruby/2.6.0/gems/ddtrace-1.4.0/ext/ddtrace_profiling_native_extension
/opt/jruby/bin/jruby -I /opt/jruby/lib/ruby/stdlib -r
./siteconf20220826-1144-1jd1na2.rb extconf.rb
+------------------------------------------------------------------------------+
| Could not compile the Datadog Continuous Profiler because                    |
| JRuby is not supported by the Datadog Continuous Profiler.                   |
|                                                                              |
| The Datadog Continuous Profiler will not be available,                       |
| but all other ddtrace features will work fine!                               |
|                                                                              |
| Get in touch with us if you're interested in profiling your app!             |
+------------------------------------------------------------------------------+
          
current directory:
/builds/honeyryderchuck/tobox/vendor/jruby/2.6.0/gems/ddtrace-1.4.0/ext/ddtrace_profiling_native_extension
make DESTDIR\= clean
sh: 1: exec: make: not found
current directory:
/builds/honeyryderchuck/tobox/vendor/jruby/2.6.0/gems/ddtrace-1.4.0/ext/ddtrace_profiling_native_extension
make DESTDIR\=
sh: 1: exec: make: not found
make failed, exit code 127
Gem files will remain installed in
/builds/honeyryderchuck/tobox/vendor/jruby/2.6.0/gems/ddtrace-1.4.0 for
inspection.
Results logged to
/builds/honeyryderchuck/tobox/vendor/jruby/2.6.0/extensions/universal-java-1.8/2.6.0/ddtrace-1.4.0/gem_make.out
```

Seems a bit weird, given that, according to the docs (and the error message), tracing should be supported in JRuby. But it's seems to be installed by default somehow?

In case the "how to" is made clear somewhere in the docs, can someone point me to it?

**Comment by ivoanjo** [comment]
Hey @HoneyryderChuck, thanks for reaching out and sorry for the issue!

TL;DR to fix this, you need to have the GNU `make` tool installed. This is usually in the `make` package on Linux/Unix package managers. After that, you will be able to use most ddtrace features in JRuby.

Hopefully this fixes the issue -- let me know if that's not the case. Also, it's great to see interest on JRuby, I'm a big fan as well (and have a few contributions too!) 😄 

---

A longer explanation of this issue is that rubygems has a mechanism for gems to compile native components on installation. On JRuby, we don't compile anything (we produce an empty file to be compiled), but rubygems still tries to compile something, and thus expects the `make` tool to be installed on the system, and calls it even without checking if it is available.

I can definitely see how this is confusing. It occurred to me that we can actually check for make being installed and give you a better error message in case it isn't, so I'll keep this ticket open until I add such a message.

**Comment by ivoanjo** [comment]
Update: Since the JRuby docker images are based on `ubuntu`, here's how to install `make`:

```bash
$ docker run --network=host -ti jruby:latest /bin/bash

# Without make it fails:
root@docker-desktop:/$ gem install ddtrace
Fetching ddtrace-1.4.0.gem
Fetching msgpack-1.5.6-java.gem
Fetching debase-ruby_core_source-0.10.16.gem
Fetching libddwaf-1.3.0.2.0.gem
Fetching libdatadog-0.7.0.1.1.gem
Successfully installed msgpack-1.5.6-java
Successfully installed libddwaf-1.3.0.2.0
Successfully installed libdatadog-0.7.0.1.1
Successfully installed debase-ruby_core_source-0.10.16
Building native extensions. This could take a while...
ERROR:  Error installing ddtrace:
	ERROR: Failed to build gem native extension.

# ... etc
sh: 1: exec: make: not found

make failed, exit code 127

root@docker-desktop:/$ apt update && apt install make
Get:1 http://archive.ubuntu.com/ubuntu focal InRelease [265 kB]
Get:2 http://security.ubuntu.com/ubuntu focal-security InRelease [114 kB]
Get:3 http://archive.ubuntu.com/ubuntu focal-updates InRelease [114 kB]
Get:4 http://archive.ubuntu.com/ubuntu focal-backports InRelease [108 kB]
Get:5 http://security.ubuntu.com/ubuntu focal-security/multiverse amd64 Packages [27.5 kB]
Get:6 http://security.ubuntu.com/ubuntu focal-security/restricted amd64 Packages [1,472 kB]
Get:7 http://archive.ubuntu.com/ubuntu focal/universe amd64 Packages [11.3 MB]
Get:8 http://security.ubuntu.com/ubuntu focal-security/universe amd64 Packages [889 kB]
Get:9 http://security.ubuntu.com/ubuntu focal-security/main amd64 Packages [2,099 kB]
Get:10 http://archive.ubuntu.com/ubuntu focal/main amd64 Packages [1,275 kB]
Get:11 http://archive.ubuntu.com/ubuntu focal/multiverse amd64 Packages [177 kB]
Get:12 http://archive.ubuntu.com/ubuntu focal/restricted amd64 Packages [33.4 kB]
Get:13 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 Packages [2,555 kB]
Get:14 http://archive.ubuntu.com/ubuntu focal-updates/universe amd64 Packages [1,183 kB]
Get:15 http://archive.ubuntu.com/ubuntu focal-updates/restricted amd64 Packages [1,583 kB]
Get:16 http://archive.ubuntu.com/ubuntu focal-updates/multiverse amd64 Packages [30.2 kB]
Get:17 http://archive.ubuntu.com/ubuntu focal-backports/universe amd64 Packages [27.4 kB]
Get:18 http://archive.ubuntu.com/ubuntu focal-backports/main amd64 Packages [55.1 kB]
Fetched 23.4 MB in 15s (1,581 kB/s)
Reading package lists... Done
Building dependency tree
Reading state information... Done
3 packages can be upgraded. Run 'apt list --upgradable' to see them.
Reading package lists... Done
Building dependency tree
Reading state information... Done
Suggested packages:
  make-doc
The following NEW packages will be installed:
  make
0 upgraded, 1 newly installed, 0 to remove and 3 not upgraded.
Need to get 162 kB of archives.
After this operation, 393 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu focal/main amd64 make amd64 4.2.1-1.2 [162 kB]
Fetched 162 kB in 0s (652 kB/s)
debconf: delaying package configuration, since apt-utils is not installed
Selecting previously unselected package make.
(Reading database ... 8921 files and directories currently installed.)
Preparing to unpack .../make_4.2.1-1.2_amd64.deb ...
Unpacking make (4.2.1-1.2) ...
Setting up make (4.2.1-1.2) ...

# Works! 😄
root@docker-desktop:/$ gem install ddtrace
Building native extensions. This could take a while...
Successfully installed ddtrace-1.4.0
1 gem installed
```

**Comment by HoneyryderChuck** [comment]
Ok, I see what you mean, that makes sense.

I'm assuming `ddtrace` doesn't have native extensions, but some of the transitive dependencies has, and probably FFI, which has been a cornercase for JRuby for a long time where rubygems is finding a hard time supporting prebuilt binaries **and** JRuby, so it resorts to compiling C extensions, thereby forcing JRuby users to install the whole C compilation toolset.

I'll close the ticket, thx for the help!

**Comment by ivoanjo** [comment]
Glad it worked!

Indeed we have two dependencies (one transitive) that need extensions -- `msgpack` and `ffi`, but both of them have JRuby variants, so you should be all set with nothing special needed for those!

**Comment by HoneyryderChuck** [comment]
Indeed, they work in JRuby, but they require "compilation on install".

While rubygems allows for releasing precompiled packages per arch for Cruby, and precompiled JARs for JRuby, it sadly does not provide "precompiled per arch for FFI" for JRuby, hence the weird pickle of having to have C compiling toolchain in a Java environment. A caveat I already know of, but was unable to identify.
